### ISSUE TITLE
Can't update Aurora cluster

### ISSUE THREAD

**Issue by jamesholmes-linktree** [pulumi/pulumi#5637]
An update to an Aurora cluster with no changes causes a panic

## Expected Behavior
I should be able to update the cluster

## Current Behavior
`[diff: ~availabilityZones]; 1 error`
`panic: fatal: An assertion has failed: Expected diff to not require deletion or replacement during Update of urn:pulumi:....`

The availability zones have not changed. They come from config yaml: 

```
  rds:availabilityZones:
  - us-west-2a
  - us-west-2b
```

## Steps to Reproduce
Create an Aurora RDS cluster. Try to update it.

## Context (Environment)
I'm trying to use Pulumi to deploy an Aurora cluster.

**Comment by clstokes** [comment]
I believe Aurora requires three availability zones for an Aurora cluster. If I create the cluster using two zones in my Pulumi application, it appears that AWS actually assigns three zones as evidenced by `aws rds describe-db-clusters`.

Pulumi certainly shouldn't panic, but I think setting three zones is the expected configuration from AWS and will unblock you on this.

#### Workaround 

Set three zones on the cluster configuration.

#### Code to reproduce

```ts
import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";

const defaultCluster = new aws.rds.Cluster("default", {
    availabilityZones: [
        "us-west-2a",
        "us-west-2b",
        // "us-west-2c",
    ],
    backupRetentionPeriod: 5,
    clusterIdentifier: "aurora-cluster-demo",
    databaseName: "mydb",
    engine: "aurora-mysql",
    masterPassword: "cameroncameron",
    masterUsername: "a0sdj121adfa",
    preferredBackupWindow: "07:00-09:00",
});
```

#### Steps to reproduce

1. `pulumi up` the code above
1. `pulumi pre` - will panic
1. uncomment the third zone
1. `pulumi up` - will complete successfully.

**Comment by jamesholmes-linktree** [comment]
Thanks for the info.

**Comment by jamesholmes-linktree** [comment]
I confirmed that CDK (and therefore CF) handles two availability zones without issue.

**Comment by clstokes** [comment]
Interesting. I tried a very minimal cloudformation template for Aurora (https://gist.github.com/clstokes/b93311e84ede345428e5d6bcdff1a0a4) and still see three zones being used even when I specify only two.

@jamesholmes-linktree if you do a `aws rds describe-db-clusters` do you see only two zones on your CDK-created cluster?

**Comment by jamesholmes-linktree** [comment]
While the cluster does indeed end up deployed with all three AZ's, 

```
"AvailabilityZones": [
    "us-west-2c",
    "us-west-2b",
    "us-west-2a"
],
```

as I mentioned above, CF handles this fine and allows updates, even if only two of these AZs are specified.

**Comment by komalali** [comment]
Looks like this worked fine (i.e. pulumi did not panic) up until pulumi-aws v3.8.0

**Comment by komalali** [comment]
Seems to be related to the preview functionality introduced in v3.9.0, passing in `PULUMI_DISABLE_PROVIDER_PREVIEW=1 pulumi pre` succeeds.

**Comment by komalali** [comment]
I'm working on fixing the underlying panic here, but I just want to reiterate that if there are only 2 availability zones specified, pulumi will replace the cluster on update due to the mismatch (the cluster is deployed with 3 AZs and you are updating with 2). This is covered in [our docs](https://www.pulumi.com/docs/reference/pkg/aws/rds/cluster/#availabilityzones_nodejs), the only way to avoid replacement is to either specify 3 availability zones or use `ignoreChanges` on this property.

**Comment by jamesholmes-linktree** [comment]
Yes, this is one of the many reasons I've swapped to CDK.

**Comment by komalali** [comment]
@jamesholmes-linktree Noted! I'd love to get more feedback from you about the other reasons for your switch if you'd be willing to share. Feel free to reach me at komal at pulumi dot com or via [twitter](https://twitter.com/zwitkali).

**Comment by joeduffy** [comment]
I dug into this a bit. I know why this is happening but I'm not sure what the intended behavior is.

The RDS cluster resource _does_ accept 2 availability zones as input for the initial provisioning. Per the above discussion, Amazon automatically assigns a third availability zone. As such, the RDS cluster resource ends up with these inputs

```json
"availabilityZones": [
    "us-west-2a",
    "us-west-2b"
],
```

and these outputs

```json
"availabilityZones": [
    "us-west-2c",
    "us-west-2b",
    "us-west-2a"
],
````

At this point, the RDS cluster _does_ have three availability zones assigned, although our program is only supplying two of them. Next time you run a `pulumi up` with no changes, however, the aforementioned panic will occur, `An assertion has failed: Expected diff to not require deletion or replacement during Update`.

The cause of the panic is as follows:

* The RDS cluster property `availabilityZones`, if changed, forces a new replacement.

* [`makeDetailedDiff` visits new properties, old properties, and then new properties again](https://github.com/pulumi/pulumi-terraform-bridge/blob/master/pkg/tfbridge/diff.go#L274-L287).

    - The `availabilityZones` value in `olds` is `[{us-west-2c} {us-west-2b} {us-west-2a}]`
    - The `availabilityZones` value in `news` is `[{us-west-2a} {us-west-2b}]`

* During the first traversal, of `olds`, the diff of the name `availability_zones.2050015877` (`us-west-2c`), at path `availabilityZones[0]`, is [correctly marked as `DELETE_REPLACE`](https://github.com/pulumi/pulumi-terraform-bridge/blob/16a63542765144a20000585f3594d022e7c14ec1/pkg/tfbridge/diff.go#L201), as it has been removed from `news` (the deletion of `us-west-2c`). This [stores `DELETE_REPLACE` into `diff["availabilityZones[0]"]`](https://github.com/pulumi/pulumi-terraform-bridge/blob/16a63542765144a20000585f3594d022e7c14ec1/pkg/tfbridge/diff.go#L218).

* During the second traversal, of `news`, however, the name at path `availabilityZones[0]` has changed. Therefore, it looks up the diff for `availability_zones.2487133097` (`us-west-2a`), which is unchanged, overwriting the above `DELETE_REPLACE` by setting `diff["availabilityZones[0]"]` to `UPDATE`. This effectively loses the fact that a replacement is needed!

A "simple" fix would be to check for duplicate `diff[path]` entries anytime we set that value and always take the "stronger" of the two (always prefer a replacement). That seems safe to me but there may be subtle reasons why that's wrong.

(Of course this is setting aside the annoying fact that ever RDS cluster provisioned this way will always want to be replaced upon update, but that seems unavoidable to me given the way the provider is working.)
